### DISCUSSION TITLE
Are page.meta variables visible in templates

### DISCUSSION THREAD

**Discussion by DeNelo** [squidfunk/mkdocs-material#7555] [Q&A] [answered]
I have a page with frontmatter defined. There is for instance a "documentation.layer" variable. I want to put the content of the variable in the copyright.html partial. But it doesn't show.
I have the mkdocs-macros plugin installed, which lets me use jinja2 syntax on the page, and I can display the page.meta variable using {{ page.meta }} - it exists. But not in the footer (which uses the copyright partial). What am I not getting ..?

If, in the template, I use page.meta.documentation.layer, it breaks on unknown keyword 'meta'. If I just say page.documentation.layer, it does not break, but just skips it (I'm testing on page.documentation). No error, but no variable ...

Now I tried using {% if page.meta['documentation'] %}, but I still get "jinja2.exceptions.UndefinedError: 'None' has no attribute 'meta'"

**Comment by alexvoss** [comment]
The templates and partials are Jinja2 template files that are processed after the Markdown is processed into HTML, after the macros plugin has done its work. So, this has nothing to do with macros. It just happens that the Macros plugin uses the Jinja2 syntax. 

I am not entirely sure why your attempt to use `page.meta` fails as this should be available in the copyright partial. The header partial, for example, uses `{{ page.meta.title }}`.

You could help us investigate the problem by creating a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Reply by DeNelo** [reply]
Thanks! 
I can't run the -info plugin because I can't build my minimum installation ... :) But  I've zipped my docs, partials, and mkdocs.yml - I hope you can see what's wrong...
[example.zip](https://github.com/user-attachments/files/17096352/example.zip)
I am still getting
`  File "C:\Users\NielsFanøe\overrides\partials\footer.html", line 51, in top-level template code
    {{ page.meta.documentation.layer }}
^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\NielsFanøe\venv\Lib\site-packages\jinja2\environment.py", line 487, in getattr
    return getattr(obj, attribute)
           ^^^^^^^^^^^^^^^^^^^^^^^
jinja2.exceptions.UndefinedError: 'None' has no attribute 'meta'`

**Reply by kamilkrzyskow** [reply, marked as answer]
MkDocs processes everything using the same templates, including the static template 404 page that doesn't include its own `page.meta`.

Therefore before every `page.meta` access you need to validate that it exists:
https://github.com/squidfunk/mkdocs-material/blob/06fe18af7ee18ed3997a7adba9398c91226d8839/src/templates/base.html#L35

> The header partial, for example, uses {{ page.meta.title }}

The header title also does this validation:
- `{% if page.meta and page.meta.title %}`

> I can't run the -info plugin because I can't build my minimum installation

iirc the `info` plugin creates the archive in `on_config` event and then quits early before the template build stage.
Therefore I believe you haven't even tried to use it :c

---

EDIT: There is also a reference
- https://squidfunk.github.io/mkdocs-material/reference/#on-a-single-page

**Reply by DeNelo** [reply]
I see - I must test for the page.meta object. I saw that on the page you referenced, but I wasn't aware that it was required.
Thanks for your help!
(And about the info plugin "not working" - I'm embarrassed - forgot that autosave was not enabled in my editor ... of course it works).
